### PR TITLE
feat: click outline and jump to relevant source/frame position

### PR DESCRIPTION
**Pull request by Myriad-Dreamin**
Click outline and jump to relevant source/frame position.

The outline jump can retrieve a span and a frame position from the heading (outline) element, but we cannot simply use the span since it may (sometime almost) dive into the package files. Hence the outline jump starts from find a suitable span in the frame by position.

Let's consider a algorithm to find a such span by position.
First, if there is a spanned frame item hit by the position, we use the frame item.
Otherwise, we perform a fuzzy search. Before search we've collected all line numbers $L = l_i$ and byte offset $o_i$ of the spanned frame item.
1. locate the first clustered line number $l = \min S$, where recusively $`S = \left\{ l_n \right\} \cup \left\{ l_i| \mid l_i - \min S|
 \leqslant \Delta\right\}`$, and $\Delta = \overline{\delta_L} + 2 \sigma(\delta_L)$.
2. find the span having the minimum $o_i$ at the line $l$

### CODE REVIEW

**Comment by Enter-tainer** [comment]
Is it related to https://github.com/Enter-tainer/typst-preview/issues/149 ? Can you also explain the notation you used?

**Comment by Myriad-Dreamin** [comment]
It is only for click action from the outline, not #149 in this PR, though we can extend it to #149. 

For notation, it uses some commonly used notations, you can also read the source code to understand it:
+ $a = min S$, where $S$ is a set, then it yields the minimum element in the set.
+ recursively $S$, where $S$ is a set definition, stops until no related elements meeting condition.
+ $\bar x$, average of a sequence.
+ $\sigma x$, standard deviation of a sequence.
+ $\delta _ L$, differential sequence of a ordered sequence.

https://github.com/Enter-tainer/typst-preview/pull/193/commits/75e1ef690088e1d55675f65c9f2782c33615ac2e#diff-fa09118099a92a2b835471862bf180af88b94f579177914d67ed16c71ba134d0R226-R262

File: src/actor/debug_loc.rs
[the remark this answers is not in the document]
Comment (Enter-tainer) [reply src/actor/debug_loc.rs:1]: I also prefer move this file and output.rs to a new dir because they are not "actors"

File: src/actor/typst.rs [hunk 1]
```diff
@@ -177,21 +180,99 @@ impl TypstClient {
...
+                self.doc2src(id).await;
+            }
+            TypstActorRequest::JumpToSrcByPosition(pos) => {
+                debug!("TypstActor: processing jump to src by position: {pos:?}");
 
+                // todo: non-main file
                 let res = self
                     .inner()
-                    .resolve_doc_to_src_jump(id)
+                    .steal_async(move |this, _| {
```
Comment (Enter-tainer) [inline src/actor/typst.rs:191]: maybe move this single huge closure to a function

File: src/actor/typst.rs [hunk 2]
```diff
@@ -177,21 +180,99 @@ impl TypstClient {
         match mail {
             TypstActorRequest::DocToSrcJumpResolve(id) => {
                 debug!("TypstActor: processing doc2src: {:?}", id);
+                self.doc2src(id).await;
+            }
+            TypstActorRequest::JumpToSrcByPosition(pos) => {
```
Comment (Enter-tainer) [inline src/actor/typst.rs:185]: When will `TypstActorRequest::JumpToSrcByPosition` used? I feel like it is not that fast

File: src/actor/typst.rs [hunk 3]
```diff
@@ -291,3 +380,93 @@ impl TypstClient {
         self.inner().add_memory_changes(MemoryEvent::Update(files))
     }
 }
+
+/// Find the position of a span in a frame.
+fn fuzzy_find_in_frame(frame: &Frame, filter: &mut impl FnMut(&Point, &Span) -> bool) -> bool {
```
Comment (Enter-tainer) [inline src/actor/typst.rs:385]: Is the return value used or not?

Comment (Enter-tainer) [reply src/actor/typst.rs:385]: We have a stateful ...filter

**Pull request closed without merging**